File: ports/ruby/lib/unicode_ruby/utf8.rb
# frozen_string_literal: true

require_relative "strict"

module UnicodeRuby
  # Strict UTF-8 codec — validator, decoder, and encoder.
  #
  # The accepted byte set is exactly the strict RFC 3629 acceptance language:
  # it rejects overlong encodings, surrogate codepoints (U+D800..U+DFFF),
  # codepoints beyond U+10FFFF, truncated multi-byte sequences, invalid start
  # bytes, and invalid continuation bytes.  Bytes are represented as arrays of
  # integers in 0..255.
  module Utf8
    module_function

    # Decoder states:
    #   :start                          — start-of-codepoint state
    #   [:cont, remaining, accum, min_cp] — open multi-byte sequence
    START = :start

    # Process one byte given the current state.  Returns one of:
    #   [:continue, new_state]
    #   [:emit, cp, new_state]
    #   [:reject, kind]
    def decode_step(state, byte)
      n = byte
      if state == START
        if n < 0x80
          [:emit, n, START]
        elsif n < 0xC2
          [:reject, Strict::INVALID_START_BYTE]
        elsif n < 0xE0
          [:continue, [:cont, 1, n & 0x1F, 0x80]]
        elsif n < 0xF0
          [:continue, [:cont, 2, n & 0x0F, 0x800]]
        elsif n < 0xF5
          [:continue, [:cont, 3, n & 0x07, 0x10000]]
        else
          [:reject, Strict::INVALID_START_BYTE]
        end
      else
        _tag, remaining, accum, min_cp = state
        return [:reject, Strict::INVALID_CONTINUATION_BYTE] if n < 0x80 || n >= 0xC0

        nxt = (accum << 6) | (n & 0x3F)
        if remaining == 1
          if nxt < min_cp
            [:reject, Strict::OVERLONG_ENCODING]
          elsif nxt >= 0xD800 && nxt <= 0xDFFF
            [:reject, Strict::SURROGATE_CODEPOINT]
          elsif nxt > 0x10FFFF
            [:reject, Strict::CODEPOINT_BEYOND_MAX]
          else
            [:emit, nxt, START]
          end
        else
          [:continue, [:cont, remaining - 1, nxt, min_cp]]
        end
      end
    end

    # The first byte offset at which the strict UTF-8 state machine rejects, as
    # `[offset, kind]`, or `nil` when the entire input is valid UTF-8.  For
    # OverlongEncoding the offset is the start byte of the offending sequence.
    # TruncatedSequence reports an offset equal to the input length.
    def first_invalid_offset(bytes)
      state = START
      seq_start = 0
      bytes.each_with_index do |b, i|
        seq_start = i if state == START
        result = decode_step(state, b)
        case result[0]
        when :continue
          state = result[1]
        when :emit
          state = result[2]
        when :reject
          kind = result[1]
          return [seq_start, kind] if kind == Strict::OVERLONG_ENCODING

          return [i, kind]
        else
          raise "unreachable utf8 step tag: #{result[0].inspect}"
        end
      end
      case state
      when START
        nil
      else
        [bytes.length, Strict::TRUNCATED_SEQUENCE]
      end
    end

    # Whole-input validity predicate.
    def valid?(bytes)
      first_invalid_offset(bytes).nil?
    end

    # Decode a UTF-8 byte array to a codepoint array.  On malformed input the
    # walker yields the longest valid prefix and stops; callers that need
    # explicit failure propagation validate first via first_invalid_offset.
    def decode_to_codepoints(bytes)
      out = []
      state = START
      bytes.each do |b|
        result = decode_step(state, b)
        case result[0]
        when :continue
          state = result[1]
        when :emit
          out << result[1]
          state = result[2]
        when :reject
          return out
        else
          raise "unreachable utf8 step tag: #{result[0].inspect}"
        end
      end
      out
    end

    # Encode a single codepoint as a 1–4 byte UTF-8 sequence.
    def encode_codepoint(cp)
      if cp < 0x80
        [cp]
      elsif cp < 0x800
        [0xC0 | (cp >> 6), 0x80 | (cp & 0x3F)]
      elsif cp < 0x10000
        [0xE0 | (cp >> 12), 0x80 | ((cp >> 6) & 0x3F), 0x80 | (cp & 0x3F)]
      else
        [0xF0 | (cp >> 18), 0x80 | ((cp >> 12) & 0x3F),
         0x80 | ((cp >> 6) & 0x3F), 0x80 | (cp & 0x3F)]
      end
    end

    # Concatenate the UTF-8 encodings of a codepoint array.
    def encode_codepoints(cps)
      out = []
      cps.each { |cp| out.concat(encode_codepoint(cp)) }
      out
    end
  end
end

File: ports/php/src/Utf8.php
<?php

declare(strict_types=1);

namespace UnicodePhp;

/// Strict UTF-8 codec — validator and decoder.
///
/// The accepted byte set is exactly the strict RFC 3629 acceptance language:
/// it rejects overlong encodings, surrogate codepoints (U+D800..U+DFFF),
/// codepoints beyond U+10FFFF, truncated multi-byte sequences, invalid start
/// bytes, and invalid continuation bytes. The accepted byte set is closed-form
/// per the spec.
///
/// Offset convention for `firstInvalidOffset`: the returned offset is the index
/// of the byte on which the state machine transitions to reject. For
/// `OverlongEncoding` (detected on emission of a multi-byte sequence) the offset
/// is the start byte of the sequence, not the last byte consumed.
final class Utf8
{
    /// The first byte offset at which the strict UTF-8 state machine rejects,
    /// or `null` when the entire input is valid UTF-8. Returns a pair
    /// `[offset, Utf8RejectKind]`. `TruncatedSequence` reports an offset equal
    /// to the input length.
    ///
    /// @param list<int> $bytes octet values in 0..255
    /// @return array{0:int,1:Utf8RejectKind}|null
    public static function firstInvalidOffset(array $bytes): ?array
    {
        // State encoding: remaining === 0 means ExpectStart; remaining > 0 is
        // an open multi-byte sequence carrying `accum` and `minCp`.
        $remaining = 0;
        $accum = 0;
        $minCp = 0;
        $seqStart = 0;
        $count = count($bytes);

        for ($i = 0; $i < $count; $i++) {
            $b = $bytes[$i];
            if ($remaining === 0) {
                $seqStart = $i;
                $n = $b;
                if ($n < 0x80) {
                    // 1-byte ASCII, emit directly; stay in ExpectStart.
                    continue;
                }
                if ($n < 0xC2) {
                    return [$i, Utf8RejectKind::InvalidStartByte];
                }
                if ($n < 0xE0) {
                    $remaining = 1;
                    $accum = $n & 0x1F;
                    $minCp = 0x80;
                } elseif ($n < 0xF0) {
                    $remaining = 2;
                    $accum = $n & 0x0F;
                    $minCp = 0x800;
                } elseif ($n < 0xF5) {
                    $remaining = 3;
                    $accum = $n & 0x07;
                    $minCp = 0x10000;
                } else {
                    return [$i, Utf8RejectKind::InvalidStartByte];
                }
                continue;
            }

            // ExpectCont: continuation bytes must lie in 0x80..0xBF.
            $n = $b;
            if ($n < 0x80 || $n >= 0xC0) {
                return [$i, Utf8RejectKind::InvalidContinuationByte];
            }
            $next = ($accum << 6) | ($n & 0x3F);
            if ($remaining === 1) {
                if ($next < $minCp) {
                    return [$seqStart, Utf8RejectKind::OverlongEncoding];
                }
                if ($next >= 0xD800 && $next <= 0xDFFF) {
                    return [$i, Utf8RejectKind::SurrogateCodepoint];
                }
                if ($next > 0x10FFFF) {
                    return [$i, Utf8RejectKind::CodepointBeyondMax];
                }
                $remaining = 0;
            } else {
                $remaining -= 1;
                $accum = $next;
            }
        }

        if ($remaining > 0) {
            return [$count, Utf8RejectKind::TruncatedSequence];
        }
        return null;
    }

    /// Whole-input validity predicate.
    ///
    /// @param list<int> $bytes
    public static function isValid(array $bytes): bool
    {
        return self::firstInvalidOffset($bytes) === null;
    }

    /// Decode a UTF-8 byte string to a codepoint list. Semantically meaningful
    /// only when the input is valid UTF-8; on malformed input the walker yields
    /// the longest valid prefix and stops. Callers that need explicit failure
    /// propagation should validate first via `firstInvalidOffset`.
    ///
    /// @param list<int> $bytes
    /// @return list<int>
    public static function decodeToCodepoints(array $bytes): array
    {
        $out = [];
        $remaining = 0;
        $accum = 0;
        $minCp = 0;

        foreach ($bytes as $b) {
            if ($remaining === 0) {
                $n = $b;
                if ($n < 0x80) {
                    $out[] = $n;
                } elseif ($n < 0xC2) {
                    return $out;
                } elseif ($n < 0xE0) {
                    $remaining = 1;
                    $accum = $n & 0x1F;
                    $minCp = 0x80;
                } elseif ($n < 0xF0) {
                    $remaining = 2;
                    $accum = $n & 0x0F;
                    $minCp = 0x800;
                } elseif ($n < 0xF5) {
                    $remaining = 3;
                    $accum = $n & 0x07;
                    $minCp = 0x10000;
                } else {
                    return $out;
                }
                continue;
            }
            $n = $b;
            if ($n < 0x80 || $n >= 0xC0) {
                return $out;
            }
            $next = ($accum << 6) | ($n & 0x3F);
            if ($remaining === 1) {
                if ($next < $minCp || ($next >= 0xD800 && $next <= 0xDFFF) || $next > 0x10FFFF) {
                    return $out;
                }
                $out[] = $next;
                $remaining = 0;
            } else {
                $remaining -= 1;
                $accum = $next;
            }
        }
        return $out;
    }
}
